Create AI Guard onboarding page (#32586)
DataDog/documentation: content/en/security/ai_guard/onboarding.md

---
title: Get Started with AI Guard
further_reading:
- link: "https://www.datadoghq.com/blog/llm-guardrails-best-practices/"
  tag: "Blog"
  text: "LLM guardrails: Best practices for deploying LLM apps securely"
---

{{< site-region region="gov" >}}<div class="alert alert-danger">AI Guard isn't available in the {{< region-param key="dd_site_name" >}} site.</div>
{{< /site-region >}}

AI Guard helps secure your AI apps and agents in real time against prompt injection, jailbreaking, tool misuse, and sensitive data exfiltration attacks. This page describes how to set it up so you can keep your data secure against these AI-based threats.

## Setup

### Prerequisites

Before you set up AI Guard, ensure you have everything you need:

> - While AI Guard is in Preview, Datadog needs to enable a backend feature flag for each organization in the Preview. Contact [Datadog support][1] with one or more Datadog organization IDs (or organization tenant names) to enable it.

Reviewer: If you use the note, you don't need to explain about the feature flag etc.

> - While AI Guard is in Preview, Datadog needs to enable a backend feature flag for each organization in the Preview. Contact [Datadog support][1] with one or more Datadog organization IDs (or organization tenant names) to enable it.

Author: @ddvijay I went back and forth about including the Preview callout in this, since I think everyone who looks at this will be in the Preview. But Michael brings up a good point about the feature flag. Can we assume that it'll be automatically enabled when a customer joins the Preview?

- Certain setup steps require specific Datadog permissions. An admin may need to create a new role with the required permissions and assign it to you.
  - To create an application key, you need the **AI Guard Evaluate** permission.

>   - If you need to make a restricted dataset so you can [limit access to AI Guard spans](#limit-access), you need the **User Access Manage** permission.

Reviewer: ```suggestion
  - To make a restricted dataset that [limits access to AI Guard spans](#limit-access), you need the **User Access Manage** permission.
```

### Usage limits {#usage-limits}

> The AI Guard Evaluator API has the following usage limits:

Reviewer: ```suggestion
The AI Guard evaluator API has the following usage limits:
```

> - 1 billion (1,000,000,000) tokens evaluated per day.

Reviewer: do we need `(1,000,000,000)`?

- 12,000 requests per minute, per IP.

If you exceed these limits, or expect to exceed them soon, contact [Datadog support][1] to discuss possible solutions.

### Create API and application keys {#create-keys}

To use AI Guard, you need at least one API key and one application key set in your Agent services, usually using environment variables. Follow in the instructions at [API and Application Keys][2] to create both.

> When you're creating your **application key**, when you're adding [scopes][3], add the `ai_guard_evaluate` scope.

Reviewer: ```suggestion
When adding [scopes][3] for the **application key**, add the `ai_guard_evaluate` scope.
```

### Set up a Datadog Agent {#agent-setup}

Datadog SDKs use the [Datadog Agent][4] to send AI Guard data to Datadog. The Agent must be running and accessible to the SDK for you to see data in Datadog.

If you don't use the Datadog Agent, the AI Guard evaluator API still works, but you can't see AI Guard traces in Datadog.

### Create a custom retention filter {#retention-filter}

To ensure no AI Guard evaluations are dropped, create a custom [retention filter][5] for AI Guard-generated spans:
- **Retention query**: `resource_name:ai_guard`
- **Span rate**: 100%
- **Trace rate**: 100%

### Limit access to AI Guard spans {#limit-access}

{{< callout url="#" btn_hidden="true" header="false">}}

> Data Access Controls is in Limited Availability. To join so you can use this feature, contact Datadog support.

Reviewer: ```suggestion
Data Access Controls is in Limited Availability. To enroll, contact Datadog support.
```

Example:

```json
{"role":"system","content":"You are a helpful AI assistant."}
```

#### User prompt format {#user-prompt-format}

A user prompt has two fields, both mandatory:
- `role`: Must be `user`.
- `content`: A string with the content of the user prompt.

Example:

```json
{"role":"user","content":"Hello World!"}
```

#### Assistant response format {#assistant-response-format}

An assistant response with no tool calls has two fields, both mandatory:
- `role`: Must be `assistant`.
- `content`: A string with the content of the user prompt.

Example:

```json
{"role":"assistant","content":"How can I help you today?"}
```

#### Assistant response with tool call format {#assistant-response-tool-call-format}

When an LLM requests the execution of a tool call, it is set in the `tool_calls` field of an assistant message. Tool calls must have a unique ID, a name of the tool, and arguments as a string (usually a JSON-serialized object).

Example:

```json
{
  "role":"assistant",
  "content":"",
  "tool_calls": [
    {
      "id": "call_123",
      "function": {
        "name": "shell",
        "arguments": "{\"command\":\"ls\"}"
      }
    }
  ]
}
```

#### Tool output format

> When the result of a tool call is passed back to the LLM, it must be formatted as a message with role `tool`, with its output in the `content` field. It must have a `tool_call_id` field that matches the content of the previous tool call request.

Reviewer: ```suggestion
When the result of a tool call is passed back to the LLM, it must be formatted as a message with role `tool`, and its output in the `content` field. It must have a `tool_call_id` field that matches the content of the previous tool call request.
```

Example:

```json
{
  "role":"tool",
  "content":". .. README.me",
  "tool_call_id": "call_123"
  }
```

### Use an SDK to create REST API calls {#sdks}

SDK instrumentation allows you to set up and monitor AI Guard activity in real time.

{{< tabs >}}
{{% tab "Python" %}}

> Beginning with version [v3.14.0rc1][1] of dd-trace-py, a new Python SDK has been introduced. This SDK provides a streamlined interface for invoking the REST API directly from Python code. The following examples demonstrate its usage:

Reviewer: just matching the link format in the JS and Java tabs

)
```

#### Example: Evaluate a user prompt {#python-example-evaluate-user-prompt}

```py
# Check if processing the user prompt is considered safe
prompt_evaluation = client.evaluate_prompt(
    history=[
        Prompt(role="system", content="You are an AI Assistant"),
    ],
    role="user",
    content="What is the weather like today?"
)
```

The `evaluate_prompt` method accepts the following parameters:
- `history` *(optional)*: A list of `Prompt` or `ToolCall` objects representing previous prompts or tool evaluations.
- `role` *(required)*: A string specifying the role associated with the prompt.
- `content` *(required)*: The content of the prompt.

The method returns a Boolean value: `True` if the prompt is considered safe to execute, or `False` otherwise. If the REST API detects potentially dangerous content, it raises an `AIGuardAbortError`.

#### Example: Evaluate a tool call {#python-example-evaluate-tool-call}

```py
# Check if executing the shell tool is considered safe
tool_evaluation = client.evaluate_tool(
    tool_name="shell",
    tool_args={"command": "shutdown"}
)
```

In this case, the `evaluate_tool` method accepts the following parameters:

> - `history` *(optional)*: A list of `Prompt` or `ToolCall` objects representing previous prompts or tool evaluations.

Reviewer: Are optional/required generally italic in our docs?

- `tool_name` *(required)*: A string specifying the name of the tool to be invoked.

> - `tool_args` *(required)*: A dictionary containing the arguments required by the tool.

Reviewer: ```suggestion
- `tool_name` *(required)*: A string specifying the name of the tool to invoke.
- `tool_args` *(required)*: A dictionary containing the required tool arguments.
```

```py
with tracer.trace("ai_guard") as span:
    result = _call_rest_api()  # REST API call

    attributes = result["data"]["attributes"]
    span.set_tag("ai_guard.target", "tool")  # Use "prompt" if evaluating a prompt
    span.set_tag("ai_guard.tool_name", "tool_name")  # Specify the tool name if applicable
    span.set_tag("ai_guard.action", attributes["action"])

    if "reason" in attributes:
        span.set_tag("ai_guard.reason", attributes["reason"])

    # Optional metadata: tags starting with ai_guard.meta will appear in the outcome table (e.g. input prompt, tool arguments, etc.)
    span.set_tag("ai_guard.meta.prompt", "the prompt that triggered the tool execution")
```

The Python SDK handles this process automatically, eliminating the need to manually create the span.

You can use the `datadog.ai_guard.evaluations` metric to count the evaluations AI Guard performed. This metric is tagged by `action`, `blocking_enabled`, `service`, and `env`.

### Set up Datadog Monitors for alerting {#set-up-datadog-monitors}

> To create monitors for alerting at certain thresholds, you can use Datadog Monitors, which is included at no additional charge in the Datadog platform. You can monitor AI Guard evaluations with either APM traces or with metrics. For both types of monitor, you should set your alert conditions, name for the alert, and define notifications; Datadog recommends using Slack.

Reviewer: you could add a link to Monitors: https://docs.datadoghq.com/monitors/

Also, pretty sure monitoring cost money: https://www.datadoghq.com/pricing/.

You might want to remove `which is included at no additional charge in the Datadog platform` so you don't have to maintain it

---
title: Get Started with AI Guard
private: true
further_reading:
- link: "https://www.datadoghq.com/blog/llm-guardrails-best-practices/"
  tag: "Blog"
  text: "LLM guardrails: Best practices for deploying LLM apps securely"
---

{{< site-region region="gov" >}}<div class="alert alert-danger">AI Guard isn't available in the {{< region-param key="dd_site_name" >}} site.</div>
{{< /site-region >}}

AI Guard helps secure your AI apps and agents in real time against prompt injection, jailbreaking, tool misuse, and sensitive data exfiltration attacks. This page describes how to set it up so you can keep your data secure against these AI-based threats.

## Setup

### Prerequisites

Before you set up AI Guard, ensure you have everything you need:

> - While AI Guard is in Preview, Datadog needs to enable a backend feature flag for each organization in the Preview. Contact [Datadog support][1] with one or more Datadog organization IDs (or organization tenant names) to enable it.

Reviewer 2: ```suggestion
- While AI Guard is in Preview, Datadog needs to enable a backend feature flag for each organization in the Preview. Contact [Datadog support][1] with one or more Datadog organization names and regions to enable it.
```

Tweaked this, since users actually do not know about the concept of organization ID, it's usually confusing to them.
